### ISSUE TITLE
Massive performance slowdown (15x) sometime between 0.20.4 and 0.20.10

### ISSUE THREAD

**Issue by veqryn** [godror/godror#113]
**Describe the bug**
Our repo has been using 0.20.4 for some time now without issue.
We have a particular query that retrieves about 7 million rows, with each row containing 3 fields (2x uint64, 1x string less than 256 characters).
We scan those rows one by one into a `map[uint64]map[string]uint64`.
The function takes about 100 seconds when run in our build environment.
After updating to 0.20.10, it started taking 1500-2000 seconds, which is more than 15x slower.
We reverted back to 0.20.4 and that fixed the issue for us. No other changes to the repo.

**To Reproduce**
Cut down version of the query code we are running
```golang
	query = squirrel.StatementBuilder.PlaceholderFormat(squirrel.Dollar).
		Select(`
			customer_id
			, email
			, email_id`).
		From("mytable")
	sqlStr, args, err = query.ToSql()
	if err != nil {
		return nil, errors.WithMessage(err, "Unable to convert query to sql")
	}

	rows, err := db.conn.QueryxContext(ctx, sqlStr, args...)
	if err != nil {
		return nil, err
	}
	defer func() { _ = rows.Close() }()

	// Manually scan and put into map
	var (
		custID     uint64
		email      string
		emailID    uint64
		custEmails = make(map[uint64]map[string]uint64)
	)
	for rows.Next() {
		if err = rows.Scan(&custID, &email, &emailID); err != nil {
			return custEmails, err
		}
		if _, ok := custEmails[custID]; !ok {
			custEmails[custID] = map[string]uint64{}
		}
		custEmails[custID][email] = emailID
	}

	return custEmails, rows.Err()
```

**Expected behavior**
No slowdown

**Your oracle client version**
12.1.0.2.0

**Your godror version**
v0.20.10

**Your go version**
1.14.12

**Your gcc version**
7.3.0 maybe?

**Machine (please complete the following information):**
OS: linux docker container for golang:1.14.12
Architecture x86_64
Version: debian buster

**Comment by tgulacsi** [comment]
What if you add `godror.FetchArraySize(8192), godror.PrefetchCount(8193)` as params to QueryContext ?

I don't see big changes since v0.20.4 - the timeout handling code has been fixed.
Another possibility is the implementation of Compose/Decompose for Number, which may confuse sql conversion (Scan) code.

**Comment by sudarshan12s** [comment]
Please try the settings suggested by @tgulacsi

The timeout handling changes introduce a handling deadline for row fetch
call. It spawns a new  go routine , polls on deadline.

stmtctx := r.statement.ctx
if stmtctx != nil {
// nil can be present when Next is issued on cursor returned from DB
done := make(chan struct{})
defer close(done)
// handle deadline for dpiStmt_fetchRows. context reused from stmt
if err := r.statement.handleDeadline(stmtctx, done); err != nil {
return err
}
}

**Comment by veqryn** [comment]
I should have time to test this coming week. Any specific godror versions I should test as well?

**Comment by tgulacsi** [comment]
c98945c7bdaae7f2087207643847f018474e3da9
7e135e6f65eb64132d60e95c91ae4ee3614d3a6a
cbd8e9e9f9da6260470fff9502cd7fce98299333
2e9dcd7ba8b71794df717bfced818c61595c8c4e

Or create a simple benchmark in a separate file and `git bisect` it.

But for querying lot of rows, increasing FetchArraySize and PrefetchCount will give significant speedup.

**Comment by tgulacsi** [comment]
# BenchmarkSelect113

    go test -run=^$ -bench=Select113 -count=1

## v0.20.10 (current master):
```
BenchmarkSelect113/simple-4                    1        2085302813 ns/op             10000 record
BenchmarkSelect113/prefetch-8192-4             8         128643942 ns/op             10000 record
```

## v0.20.4
```
BenchmarkSelect113/simple-4                    1        2058920979 ns/op             10000 record
BenchmarkSelect113/prefetch-8192-4             8         126195669 ns/op             10000 record
```

There's no difference in timing, but using Prefetch, it's 20 times faster.

**Comment by tgulacsi** [comment]
See the [Tuning Guide](https://godror.github.io/godror/doc/tuning.html#-query-performance) - you can increase the performance for large number of rows by increasing the FetchArraySize and PrefetchCount.
But this increases the memory used, and the latency.

So, this is the classical tradeoff between latency and throughput.

As the prefetch count is set at execution time, we cannot make this adaptive, thus the caller must know.

**Comment by tgulacsi** [comment]
Benchmarking the Time To First Record:
```
BenchmarkSelect113/simple-4                   13         210138583 ns/op             22714 firstRecordMicros            1300 record
BenchmarkSelect113/prefetch-2-4                1        1090893649 ns/op                82.0 firstRecordMicros           100 record
BenchmarkSelect113/prefetch-16-4               6         507805542 ns/op                99.0 firstRecordMicros           600 record
BenchmarkSelect113/prefetch-128-4             24         268103036 ns/op               283 firstRecordMicros            2400 record
BenchmarkSelect113/prefetch-1024-4            13          79410532 ns/op              1006 firstRecordMicros            1300 record
PASS
```

**Comment by tgulacsi** [comment]
Reporting the read bytes/s:
```
name                       speed
Select113/simple-4              228kB/s ±21%
Select113/prefetch-1-4          0.00B/s
Select113/prefetch-2-4          0.00B/s
Select113/prefetch-4-4         10.0kB/s ± 0%
Select113/prefetch-8-4         20.0kB/s ± 0%
Select113/prefetch-16-4        40.0kB/s ± 0%
Select113/prefetch-32-4        86.0kB/s ± 7%
Select113/prefetch-64-4         160kB/s ±12%
Select113/prefetch-128-4        320kB/s ± 6%
Select113/prefetch-256-4        520kB/s ±13%
Select113/prefetch-512-4        792kB/s ±15%
Select113/prefetch-1024-4       902kB/s ±27%
Select113/prefetch-2048-4       878kB/s ±40%
Select113/prefetch-4096-4       396kB/s ± 7%
Select113/prefetch-8192-4       206kB/s ±27%

name                       firstRecordMicros
Select113/simple-4                18.7k ± 2%
Select113/prefetch-1-4             31.2 ±60%
Select113/prefetch-2-4             48.0 ±48%
Select113/prefetch-4-4             44.2 ±47%
Select113/prefetch-8-4             48.8 ±62%
Select113/prefetch-16-4            69.6 ±81%
Select113/prefetch-32-4            68.2 ±96%
Select113/prefetch-64-4           85.6 ±154%
Select113/prefetch-128-4            242 ±70%
Select113/prefetch-256-4            351 ±92%
Select113/prefetch-512-4            825 ±98%
Select113/prefetch-1024-4         3.91k ±48%
Select113/prefetch-2048-4         4.74k ±78%
Select113/prefetch-4096-4         7.92k ±86%
Select113/prefetch-8192-4         8.00k ±25%
```

shows that 1024 bytes of prefetch gives the highest throughput, 
but 3910ms as time to first record - for latency, 2 rows of prefetch gives the absolute best.

**Comment by cjbj** [comment]
Prefetch should help the server scale, since it will need fewer interactions with the client.  

If the time for first record is important, don't forget the SQL side, e.g. [FIRST_ROWS(n)](https://asktom.oracle.com/pls/apex/f?p=100:11:0::::P11_QUESTION_ID:9537748500346205895).

**Comment by veqryn** [comment]
Ok, so I raw some new tests today, and this is the result:

We run all our tests with the `-race` flag, to hopefully catch any race conditions. We also have a single canary pod up that is running the full application with the race flag as well, to detect races in production.

Using v0.20.4 of godror, we are able to scan in all 7 million rows from a table in our test. The time taken for the 3 times I ran the test today:
113s
69s
73s

After upgrading to v0.20.10 of godror, we saw the following:
2386s
1811s

After adding the larger fetch array size and prefetch size, with v0.20.10 of godror:
1598s
1732s

Then I figured I'd try running the tests without the race flag.
Turns out they pass just fine (less than 60 seconds) for v0.20.10 of godror (both with and without fetch options).

So I guess the problem is that something between v0.20.4 and v0.20.10 interacts very badly with the `-race` flag.

I'd still consider this a major bug, as we run with the -race flag in production.

**Comment by tgulacsi** [comment]
I don't know what does the race - my experience that it slows down the program substantially, and balloons memory usage exponentially. So I only use it for testing.

TL;DR; if you want performance, don't use -race !

If you're curious, you can `git bisect` with `git bisect good v0.20.4`, `git bisect bad v0.20.4` and then `git bisect run test.sh`
where `test.sh` should run some performance test (`go run -race spec_test.go`), and exit with 0 for good (fast) result and with 1 if this is slow version. (Should exit with 125 if this version cannot be tested - for example cannot be compiled).

With `git bisect run` you can automate the whole process. But as we only have 29 commits between v0.20.4 and v0.20.10, the bisect needs 5 runs, so if the test is not too slow (say, break it when it's longer than 2 minutes), then you can find the commit in 10-15 minutes.

**Comment by veqryn** [comment]
It was a bit difficult to run due to golang modules and vendoring, and having to do it in our build environment instead of locally, but in the end this is what I got:
`7e135e6f65eb64132d60e95c91ae4ee3614d3a6a is the first bad commit`

**Comment by tgulacsi** [comment]
Thanks for the bisect!

Strange, I cannot reproduce the difference with `go test -race -count=1 -run=^$ -bench=Select113`

I've compared v0.20.4, c98945c (the commit before 7e135e6), 7e135e6 (the one you found bad), and the current master:

```
$ benchstat -norange /tmp/113-v0.20.4.txt /tmp/113-c98945c.txt /tmp/113-7e135e6.txt /tmp/113-5dc4877.txt
name \ time/op             /tmp/113-v0.20.4.txt  /tmp/113-c98945c.txt  /tmp/113-7e135e6.txt  /tmp/113-5dc4877.txt
Select113/simple-4                   199ms ± 0%            135ms ± 0%            200ms ± 0%            210ms ± 0%
Select113/prefetch-1-4               1.82s ± 0%            1.80s ± 0%            1.91s ± 0%            2.04s ± 0%
Select113/prefetch-2-4               1.40s ± 0%            1.38s ± 0%            1.45s ± 0%            1.05s ± 0%
Select113/prefetch-4-4               722ms ± 0%            716ms ± 0%            739ms ± 0%            801ms ± 0%
Select113/prefetch-8-4               490ms ± 0%            492ms ± 0%            507ms ± 0%            550ms ± 0%
Select113/prefetch-16-4              499ms ± 0%            512ms ± 0%            462ms ± 0%            499ms ± 0%
Select113/prefetch-32-4              364ms ± 0%            356ms ± 0%            376ms ± 0%            375ms ± 0%
Select113/prefetch-64-4              332ms ± 0%            328ms ± 0%            350ms ± 0%            326ms ± 0%
Select113/prefetch-128-4             253ms ± 0%            256ms ± 0%            267ms ± 0%            263ms ± 0%
Select113/prefetch-256-4             169ms ± 0%            164ms ± 0%            159ms ± 0%            165ms ± 0%
Select113/prefetch-512-4             112ms ± 0%             88ms ± 0%            111ms ± 0%            116ms ± 0%
Select113/prefetch-1024-4           84.9ms ± 0%           91.1ms ± 0%           84.9ms ± 0%           84.3ms ± 0%
Select113/prefetch-2048-4           71.9ms ± 0%           73.6ms ± 0%           75.5ms ± 0%           72.6ms ± 0%
Select113/prefetch-4096-4           92.1ms ± 0%           91.9ms ± 0%           99.1ms ± 0%          100.3ms ± 0%
Select113/prefetch-8192-4            116ms ± 0%            113ms ± 0%            121ms ± 0%            120ms ± 0%

name \ speed               /tmp/113-v0.20.4.txt  /tmp/113-c98945c.txt  /tmp/113-7e135e6.txt  /tmp/113-5dc4877.txt
Select113/simple-4                 230kB/s ± 0%          190kB/s ± 0%          210kB/s ± 0%          200kB/s ± 0%
Select113/prefetch-1-4             0.00B/s               0.00B/s               0.00B/s               0.00B/s
Select113/prefetch-2-4             0.00B/s               0.00B/s               0.00B/s               0.00B/s
Select113/prefetch-4-4            10.0kB/s ± 0%         10.0kB/s ± 0%         10.0kB/s ± 0%         10.0kB/s ± 0%
Select113/prefetch-8-4            20.0kB/s ± 0%         20.0kB/s ± 0%         20.0kB/s ± 0%         20.0kB/s ± 0%
Select113/prefetch-16-4           50.0kB/s ± 0%         40.0kB/s ± 0%         40.0kB/s ± 0%         40.0kB/s ± 0%
Select113/prefetch-32-4           90.0kB/s ± 0%         90.0kB/s ± 0%         90.0kB/s ± 0%         80.0kB/s ± 0%
Select113/prefetch-64-4            180kB/s ± 0%          180kB/s ± 0%          170kB/s ± 0%          150kB/s ± 0%
Select113/prefetch-128-4           310kB/s ± 0%          320kB/s ± 0%          290kB/s ± 0%          270kB/s ± 0%
Select113/prefetch-256-4           500kB/s ± 0%          480kB/s ± 0%          450kB/s ± 0%          410kB/s ± 0%
Select113/prefetch-512-4           700kB/s ± 0%          590kB/s ± 0%          650kB/s ± 0%          590kB/s ± 0%
Select113/prefetch-1024-4          800kB/s ± 0%          750kB/s ± 0%          760kB/s ± 0%          620kB/s ± 0%
Select113/prefetch-2048-4          630kB/s ± 0%          620kB/s ± 0%          600kB/s ± 0%          630kB/s ± 0%
Select113/prefetch-4096-4          460kB/s ± 0%          420kB/s ± 0%          430kB/s ± 0%          320kB/s ± 0%
Select113/prefetch-8192-4          250kB/s ± 0%          260kB/s ± 0%          240kB/s ± 0%          240kB/s ± 0%

name \ firstRecordMicros   /tmp/113-v0.20.4.txt  /tmp/113-c98945c.txt  /tmp/113-7e135e6.txt  /tmp/113-5dc4877.txt
Select113/simple-4                   18.9k ± 0%            18.5k ± 0%            20.0k ± 0%            22.1k ± 0%
Select113/prefetch-1-4                 164 ± 0%               83 ± 0%              104 ± 0%              167 ± 0%
Select113/prefetch-2-4                 152 ± 0%              135 ± 0%              229 ± 0%              160 ± 0%
Select113/prefetch-4-4                 203 ± 0%              126 ± 0%              137 ± 0%              240 ± 0%
Select113/prefetch-8-4                 177 ± 0%              161 ± 0%               85 ± 0%              198 ± 0%
Select113/prefetch-16-4               52.0 ± 0%            220.0 ± 0%            180.0 ± 0%            182.0 ± 0%
Select113/prefetch-32-4                328 ± 0%              232 ± 0%              391 ± 0%              213 ± 0%
Select113/prefetch-64-4                207 ± 0%              123 ± 0%              329 ± 0%              307 ± 0%
Select113/prefetch-128-4               304 ± 0%              517 ± 0%              598 ± 0%              375 ± 0%
Select113/prefetch-256-4               161 ± 0%              932 ± 0%              840 ± 0%              563 ± 0%
Select113/prefetch-512-4               602 ± 0%             1499 ± 0%             1390 ± 0%             1506 ± 0%
Select113/prefetch-1024-4            2.65k ± 0%            2.74k ± 0%            3.24k ± 0%            3.87k ± 0%
Select113/prefetch-2048-4            5.47k ± 0%            5.29k ± 0%            5.09k ± 0%            1.88k ± 0%
Select113/prefetch-4096-4            11.4k ± 0%            11.6k ± 0%             9.7k ± 0%            10.6k ± 0%
Select113/prefetch-8192-4            17.4k ± 0%            20.2k ± 0%            18.7k ± 0%            22.0k ± 0%
```

Maybe BenchmarkSelect113 in `z_bench_113_test.go` does not measure what you measure - can you tweak it to resemble your test case? Or show us your measuring code?

Another possibility is the difference between the environment - I've tested this on linux/amd64 (5.9), Debian bullseye/sid, Go 1.15.6.

**Comment by sudarshan12s** [comment]
I suspect the below change of that commit might be causing an issue since
it checks deadline for fetch calls and involves a seperate goroutine for
that . I am doing more tests, will update.

stmtctx := r.statement.ctx if stmtctx != nil { // nil can be present when
Next is issued on cursor returned from DB done := make(chan struct{}) defer
close(done) // handle deadline for dpiStmt_fetchRows. context reused from
stmt if err := r.statement.handleDeadline(stmtctx, done); err != nil {
return err } }

**Comment by tgulacsi** [comment]
@sudarshan12s that could the cause, yes. An alleviation could be to have one watchdog for all rows.Next calls - I'll try this in a minute.

@veqryn can you test on a "bad" version (can be the latest, or v0.20.10) with commenting out the incriminated lines in rows.go (see https://github.com/godror/godror/issues/113#issuecomment-742264918) ?

**Comment by tgulacsi** [comment]
I don't see big differences between the "bad" version vs. cc0f47750e9d132f4c530888465135db9a388aaa vs. the no-watchdog-at-all versions:

```
$ benchstat /tmp/113-wd_each.txt /tmp/113-wd_once.txt /tmp/113-wd_none.txt
name \ time/op             /tmp/113-wd_each.txt  /tmp/113-wd_once.txt  /tmp/113-wd_none.txt
Select113/simple-8                   230ms ± 0%            359ms ± 1%            338ms ± 9%
Select113/prefetch-1-8               2.50s ± 0%            2.48s ± 1%            2.46s ± 2%
Select113/prefetch-2-8               1.23s ± 0%            1.27s ± 1%            1.24s ± 1%
Select113/prefetch-4-8               944ms ± 0%            945ms ± 2%            946ms ± 1%
Select113/prefetch-8-8               627ms ± 0%            641ms ± 2%            643ms ± 2%
Select113/prefetch-16-8              570ms ± 0%            572ms ± 1%            569ms ± 2%
Select113/prefetch-32-8              458ms ± 0%            449ms ± 8%            454ms ± 1%
Select113/prefetch-64-8              539ms ± 0%            495ms ± 3%            502ms ± 2%
Select113/prefetch-128-8             473ms ± 0%            491ms ±17%            497ms ± 4%
Select113/prefetch-256-8             257ms ± 0%            253ms ±27%            258ms ± 7%
Select113/prefetch-512-8             104ms ± 0%            149ms ±14%            135ms ±31%
Select113/prefetch-1024-8           90.0ms ± 0%           95.4ms ± 2%           97.2ms ±16%
Select113/prefetch-2048-8           55.1ms ± 0%           61.6ms ±12%           61.6ms ±13%
Select113/prefetch-4096-8           50.1ms ± 0%           50.5ms ± 1%           63.4ms ±42%
Select113/prefetch-8192-8           67.6ms ± 0%           91.5ms ±25%           84.4ms ± 5%

name \ speed               /tmp/113-wd_each.txt  /tmp/113-wd_once.txt  /tmp/113-wd_none.txt
Select113/simple-8                 180kB/s ± 0%          190kB/s ± 0%          190kB/s ± 0%
Select113/prefetch-1-8             0.00B/s               0.00B/s               0.00B/s
Select113/prefetch-2-8             0.00B/s               0.00B/s               0.00B/s
Select113/prefetch-4-8            10.0kB/s ± 0%         10.0kB/s ± 0%         10.0kB/s ± 0%
Select113/prefetch-8-8            20.0kB/s ± 0%         16.0kB/s ±38%         10.0kB/s ± 0%
Select113/prefetch-16-8           30.0kB/s ± 0%         30.0kB/s ± 0%         30.0kB/s ± 0%
Select113/prefetch-32-8           70.0kB/s ± 0%         70.0kB/s ± 0%         70.0kB/s ± 0%
Select113/prefetch-64-8            130kB/s ± 0%          130kB/s ± 0%          130kB/s ± 0%
Select113/prefetch-128-8           260kB/s ± 0%          256kB/s ± 2%          260kB/s ± 0%
Select113/prefetch-256-8           470kB/s ± 0%          480kB/s ± 0%          478kB/s ± 3%
Select113/prefetch-512-8           740kB/s ± 0%          806kB/s ± 1%          824kB/s ± 9%
Select113/prefetch-1024-8         1.20MB/s ± 0%         1.22MB/s ± 3%         1.18MB/s ± 3%
Select113/prefetch-2048-8         1.45MB/s ± 0%         1.66MB/s ± 8%         1.62MB/s ±12%
Select113/prefetch-4096-8         1.66MB/s ± 0%         1.96MB/s ±27%         1.32MB/s ±55%
Select113/prefetch-8192-8          710kB/s ± 0%          476kB/s ±24%          516kB/s ±14%

name \ firstRecordMicros   /tmp/113-wd_each.txt  /tmp/113-wd_once.txt  /tmp/113-wd_none.txt
Select113/simple-8                   25.7k ± 0%            25.9k ± 8%            25.1k ± 3%
Select113/prefetch-1-8                48.0 ± 0%            149.6 ±53%            112.8 ±72%
Select113/prefetch-2-8                36.0 ± 0%            129.8 ±56%            128.4 ±27%
Select113/prefetch-4-8                45.0 ± 0%            124.0 ±65%             88.0 ±51%
Select113/prefetch-8-8                39.0 ± 0%            161.2 ±38%            102.6 ±60%
Select113/prefetch-16-8               53.0 ± 0%            165.2 ±52%            119.6 ±46%
Select113/prefetch-32-8               77.0 ± 0%            189.8 ± 7%            168.4 ±44%
Select113/prefetch-64-8                103 ± 0%              185 ± 1%              142 ±58%
Select113/prefetch-128-8               112 ± 0%              275 ±74%              232 ±58%
Select113/prefetch-256-8               597 ± 0%              330 ±73%              360 ±64%
Select113/prefetch-512-8             1.34k ± 0%            0.94k ±69%            0.89k ±48%
Select113/prefetch-1024-8            2.33k ± 0%            2.34k ±15%            2.15k ±20%
Select113/prefetch-2048-8            3.90k ± 0%            3.30k ±61%            3.91k ±37%
Select113/prefetch-4096-8            9.42k ± 0%            7.20k ±48%            9.38k ± 4%
Select113/prefetch-8192-8            11.2k ± 0%            18.5k ±10%            18.6k ±10%
```

Hmm, an idea: @veqryn, do you use deadlines?

**Comment by veqryn** [comment]
We do create a `context.WithTimeout(ctx, ...)` before every single database call, with an appropriate timeout.

Commenting out the lines in rows.go worked, the test passed.

**Comment by tgulacsi** [comment]
@veqryn please try 678a447 !

Some performance (thorughput) came back with it:

```
name \ time/op             /tmp/113-lng.txt  /tmp/113-bad.txt  /tmp/113-678a447.txt
Select113/simple-8               283ms ± 0%        231ms ± 0%            280ms ± 0%
Select113/prefetch-1-8           2.60s ± 0%        2.57s ± 0%            2.53s ± 0%
Select113/prefetch-2-8           1.39s ± 0%        1.31s ± 0%            1.26s ± 0%
Select113/prefetch-4-8           1.01s ± 0%        1.02s ± 0%            0.94s ± 0%
Select113/prefetch-8-8           674ms ± 0%        517ms ± 0%            626ms ± 0%
Select113/prefetch-16-8          601ms ± 0%        465ms ± 0%            575ms ± 0%
Select113/prefetch-32-8          469ms ± 0%        460ms ± 0%            544ms ± 0%
Select113/prefetch-64-8          520ms ± 0%        490ms ± 0%            508ms ± 0%
Select113/prefetch-128-8         484ms ± 0%        597ms ± 0%            470ms ± 0%
Select113/prefetch-256-8         270ms ± 0%        282ms ± 0%            250ms ± 0%
Select113/prefetch-512-8         138ms ± 0%        121ms ± 0%            165ms ± 0%
Select113/prefetch-1024-8       96.5ms ± 0%      138.0ms ± 0%           82.7ms ± 0%
Select113/prefetch-2048-8       54.0ms ± 0%       90.3ms ± 0%           55.9ms ± 0%
Select113/prefetch-4096-8       51.0ms ± 0%      129.0ms ± 0%           48.9ms ± 0%
Select113/prefetch-8192-8       77.6ms ± 0%       85.5ms ± 0%           59.4ms ± 0%

name \ speed               /tmp/113-lng.txt  /tmp/113-bad.txt  /tmp/113-678a447.txt
Select113/simple-8             180kB/s ± 0%      120kB/s ± 0%          180kB/s ± 0%
Select113/prefetch-1-8         0.00B/s           0.00B/s               0.00B/s
Select113/prefetch-2-8         0.00B/s           0.00B/s               0.00B/s
Select113/prefetch-4-8        10.0kB/s ± 0%     10.0kB/s ± 0%         10.0kB/s ± 0%
Select113/prefetch-8-8        10.0kB/s ± 0%     10.0kB/s ± 0%         20.0kB/s ± 0%
Select113/prefetch-16-8       30.0kB/s ± 0%     30.0kB/s ± 0%         30.0kB/s ± 0%
Select113/prefetch-32-8       70.0kB/s ± 0%     60.0kB/s ± 0%         70.0kB/s ± 0%
Select113/prefetch-64-8        130kB/s ± 0%      100kB/s ± 0%          130kB/s ± 0%
Select113/prefetch-128-8       260kB/s ± 0%      170kB/s ± 0%          260kB/s ± 0%
Select113/prefetch-256-8       470kB/s ± 0%      250kB/s ± 0%          470kB/s ± 0%
Select113/prefetch-512-8       790kB/s ± 0%      320kB/s ± 0%          810kB/s ± 0%
Select113/prefetch-1024-8     1.19MB/s ± 0%     0.55MB/s ± 0%         1.24MB/s ± 0%
Select113/prefetch-2048-8     1.47MB/s ± 0%     0.53MB/s ± 0%         1.55MB/s ± 0%
Select113/prefetch-4096-8     1.68MB/s ± 0%     0.76MB/s ± 0%         1.70MB/s ± 0%
Select113/prefetch-8192-8      900kB/s ± 0%      480kB/s ± 0%         1180kB/s ± 0%

name \ firstRecordMicros   /tmp/113-lng.txt  /tmp/113-bad.txt  /tmp/113-678a447.txt
Select113/simple-8               25.0k ± 0%        24.5k ± 0%            26.3k ± 0%
Select113/prefetch-1-8             125 ± 0%          105 ± 0%              266 ± 0%
Select113/prefetch-2-8            85.0 ± 0%        100.0 ± 0%            117.0 ± 0%
Select113/prefetch-4-8             124 ± 0%          371 ± 0%              376 ± 0%
Select113/prefetch-8-8             125 ± 0%          116 ± 0%              320 ± 0%
Select113/prefetch-16-8            181 ± 0%          321 ± 0%              347 ± 0%
Select113/prefetch-32-8            162 ± 0%          456 ± 0%              100 ± 0%
Select113/prefetch-64-8            290 ± 0%          405 ± 0%              446 ± 0%
Select113/prefetch-128-8           309 ± 0%          190 ± 0%              465 ± 0%
Select113/prefetch-256-8           536 ± 0%          669 ± 0%              694 ± 0%
Select113/prefetch-512-8         1.02k ± 0%        0.46k ± 0%            1.18k ± 0%
Select113/prefetch-1024-8        2.10k ± 0%        2.41k ± 0%            2.36k ± 0%
Select113/prefetch-2048-8        1.42k ± 0%        4.16k ± 0%            4.35k ± 0%
Select113/prefetch-4096-8        7.75k ± 0%        6.53k ± 0%            9.64k ± 0%
Select113/prefetch-8192-8        16.9k ± 0%         4.9k ± 0%             5.2k ± 0%
```

**Comment by veqryn** [comment]
678a447c5f24010c771a37186607457b06f02333 works for me.

I take it that the main problem was that the driver was spawning 1 goroutine every 100 rows, which meant 70,000 new goroutines in my case?  (If so, it was a bit odd that I still had problems with a fetch array size of 8192, as that would have meant just 855 new goroutines, which isn't too bad.)

Anyway, thanks for fixing this.

I'm still a bit surprised the problem doesn't show up in your latest benchmarks, unless I am reading them wrong?

**Comment by tgulacsi** [comment]
Thanks for testing it!

It has spawned a goroutine per Next, and created a channel and closed the channel and destroyed the goroutine.

Benchmarks: not a big diff, but for throughput, the first and third columns are roughly the same, the second (the first bad) is roughly half of it. So not that big difference, but there is.

**Comment by sudarshan12s** [comment]
Thanks @tgulacsi for the changes.

The idea of checking deadline in each Next call was to allow user
cancellations and complying to deadline possibly in each Go function.
missed the performance part, Thanks  @veqryn for pointing the issue .

**Comment by tgulacsi** [comment]
@sudarshan12s , it's still obeys the context with its Deadline - just only one watchdog goroutine is started, which Break()s when the context is canceled or times out, and finishes gracefully when the Rows is exhausted (io.EOF) or Close()d. 
Try it!

**Comment by sudarshan12s** [comment]
> @sudarshan12s , it's still obeys the context with its Deadline - just only one watchdog goroutine is started, which Break()s when the context is canceled or times out, and finishes gracefully when the Rows is exhausted (io.EOF) or Close()d.
> Try it!

Thanks I tried few test cases. Yes I could see rows.close would close the channel associated with goroutine spawned.
